### ISSUE TITLE
Numpy > 2 resolution doesn't seem to be working.

### ISSUE THREAD

**Issue by IAlibay** [conda-forge/gsd-feedstock#79]
### Solution to issue cannot be found in the documentation.

- [x] I checked the documentation.

### Issue

Installing gsd with numpy<2 yield gsd 3.4.1, even though the recipe states a range of "numpy>=2".

Example:

> mamba install "numpy<2" gsd

Yields:

Looking for: ["numpy[version='<2']", 'gsd']

conda-forge/linux-64                                        Using cache
conda-forge/noarch                                          Using cache

Pinned packages:
  - python 3.10.*

Transaction

  Prefix: /home/bioc1523/software/mambaforge/install/envs/test-mda

  Updating specs:

   - numpy[version='<2']
   - gsd
   - ca-certificates
   - openssl

  Package         Version  Build                Channel                    Size
─────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────

  + gsd             3.4.1  py310ha75aee5_0      conda-forge/linux-64      247kB
  + libblas         3.9.0  25_linux64_openblas  conda-forge/linux-64       16kB
  + libcblas        3.9.0  25_linux64_openblas  conda-forge/linux-64       16kB
  + libgfortran    14.2.0  h69a702a_1           conda-forge/linux-64       54kB
  + libgfortran5   14.2.0  hd5240d6_1           conda-forge/linux-64        1MB
  + liblapack       3.9.0  25_linux64_openblas  conda-forge/linux-64       16kB
  + libopenblas    0.3.28  pthreads_h94d23a6_1  conda-forge/linux-64        6MB
  + libstdcxx      14.2.0  hc0a3c3a_1           conda-forge/linux-64        4MB
  + libstdcxx-ng   14.2.0  h4852527_1           conda-forge/linux-64       54kB
  + numpy          1.26.4  py310hb13e2d6_0      conda-forge/linux-64     Cached
  + python_abi       3.10  5_cp310              conda-forge/linux-64        6kB

### Installed packages

```shell
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
bzip2                     1.0.8                h4bc722e_7    conda-forge
ca-certificates           2024.8.30            hbcca054_0    conda-forge
ld_impl_linux-64          2.43                 h712a8e2_2    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc                    14.2.0               h77fa898_1    conda-forge
libgcc-ng                 14.2.0               h69a702a_1    conda-forge
libgomp                   14.2.0               h77fa898_1    conda-forge
libnsl                    2.0.1                hd590300_0    conda-forge
libsqlite                 3.47.0               hadc24fc_1    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libxcrypt                 4.4.36               hd590300_1    conda-forge
libzlib                   1.3.1                hb9d3cd8_2    conda-forge
ncurses                   6.5                  he02047a_1    conda-forge
openssl                   3.3.2                hb9d3cd8_0    conda-forge
pip                       24.3.1             pyh8b19718_0    conda-forge
python                    3.10.15         h4a871b0_2_cpython    conda-forge
readline                  8.2                  h8228510_1    conda-forge
setuptools                75.3.0             pyhd8ed1ab_0    conda-forge
tk                        8.6.13          noxft_h4845f30_101    conda-forge
tzdata                    2024b                hc8b5060_0    conda-forge
wheel                     0.45.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
```

### Environment info

```shell
active environment : test-mda
    active env location : /home/bioc1523/software/mambaforge/install/envs/test-mda
            shell level : 1
       user config file : /home/bioc1523/.condarc
 populated config files : /home/bioc1523/software/mambaforge/install/.condarc
                          /home/bioc1523/.condarc
          conda version : 22.9.0
    conda-build version : not installed
         python version : 3.10.8.final.0
       virtual packages : __cuda=12.5=0
                          __linux=5.15.153.1=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/bioc1523/software/mambaforge/install  (writable)
      conda av data dir : /home/bioc1523/software/mambaforge/install/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/bioc1523/software/mambaforge/install/pkgs
                          /home/bioc1523/.conda/pkgs
       envs directories : /home/bioc1523/software/mambaforge/install/envs
                          /home/bioc1523/.conda/envs
               platform : linux-64
             user-agent : conda/22.9.0 requests/2.28.1 CPython/3.10.8 Linux/5.15.153.1-microsoft-standard-WSL2 ubuntu/20.04.4 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

**Comment by joaander** [comment]
The numpy 2.0 migrator removed numpy from the run requirements. In #72, it states:

> Additionally, we no longer need to use {{ pin_compatible("numpy") }} as a
run requirement - this has been handled for more than two years now by a
run-export on the numpy package itself. The migrator will therefore remove
any occurrences of this.

The root cause of your problem is that numpy 2.0 considers numpy 1.21 as a compatible version: 
https://github.com/conda-forge/numpy-feedstock/blob/1b3ce6a006b96f3356d252ac874907398a88a910/recipe/meta.yaml#L6
https://github.com/conda-forge/numpy-feedstock/blob/1b3ce6a006b96f3356d252ac874907398a88a910/recipe/meta.yaml#L50-L51

**Comment by IAlibay** [comment]
Thanks @joaander 

> The root cause of your problem is that numpy 2.0 considers numpy 1.21 as a compatible version:

That makes sense - it is runtime compatible, but it's making `pip check` fail in the oddest ways. I guess the "issue" here is that the GSD pyproject.toml has numpy > 2 as a runtime dependency, so pip check gets lost.

Now generally my take is "pip check doesn't matter anyways", but it looks like it's a default check for conda-forge recipes (or at least it's affected this case: https://github.com/conda-forge/staged-recipes/pull/28116)

**Comment by IAlibay** [comment]
> I suppose we could add numpy >= 2 to the run requirements section of the gsd recipe. I am not familiar enough with the conda-forge build ecosystem to know whether that will break things.

@tylerjereddy - sorry to ping you here, I suspect this might be something you would be more familar with. As this is a dependency "issue" for some MDAnalysis components, is this something you might be able to advise on?

**Comment by joaander** [comment]
I think this is a question for the conda-forge administrators. Their docs state to use `pin_compatible` even though that is no longer correct: https://conda-forge.org/docs/maintainer/knowledge_base/#building-against-numpy .

**Comment by tylerjereddy** [comment]
> I suppose we could add numpy >= 2 to the run requirements section of the gsd recipe. I am not familiar enough with the conda-forge build ecosystem to know whether that will break things.

It sounds like you're using a super-new NumPy feature with the new string machinery that requires NumPy > 2.0 at *runtime*. Let's check with @h-vetinari, who usually handles conda-forge matters upstream, on whether that is the correct course of action.

**Comment by h-vetinari** [comment]
Yeah, if you have newer runtime requirements than what numpy imposes by default, then you need to add that as a constraint under `run:`.
